### PR TITLE
Factor out iso selection on instance launch

### PR DESCRIPTION
**Pull request by edlerd**
## Done

- extracted dedicated list for iso files when picking an image for the instance launch

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @lorumic or @edlerd for access.
    - With a local copy of this branch, run as described [here](../HACKING.md#setting-up-for-development).
2. Perform the following QA steps:
    - create instance > browse images > use custom iso > upload custom iso
    - ensure the instance creation flow with uploading or selecting a custom iso file works flawless

### CODE REVIEW

File: src/pages/images/IsoSelector.tsx [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                primaryImage.pool === image.pool
+                  ? "positive"
+                  : ""
+              }
+              onClick={() => onSelect(image, "virtual-machine")}
+              dense
+            >
+              Select
+            </Button>
+          ),
```
Comment (lorumic) [inline src/pages/images/IsoSelector.tsx:80]: Screenshot from 2023-09-06 09:38:26

My gut feeling tells me that users would appreciate a delete button to get rid of ISOs much more here than where it is now (hidden in the storage pool detail page, volumes tab). I suggest to have it in both places: we could add just a delete icon at the end of the row, after the "Select" button.

File: src/pages/images/ImageSelectorModal.tsx [hunk 2]
```diff
@@ -14,31 +15,50 @@ export interface IsoImage {
...
+      )}
+      {content === SELECT_ISO && (
+        <IsoSelector
           primaryImage={primary}
           onSelect={onSelect}
-          onUpload={() => setContent(UPLOAD)}
+          onUpload={() => setContent(UPLOAD_ISO)}
+          onCancel={() => setContent(SELECT_IMAGE)}
         />
       )}
```
[the remark this answers is not in the document]
Comment (edlerd) [reply src/pages/images/ImageSelectorModal.tsx]: On the first interaction with an empty list, we want to make it clear the iso is available later as well and not just for the one instance. That's why we always go to the select list - before and after uploading from the instance creation.

**Comment by edlerd** [comment]
Changed the behaviour of the selection a bit, so now we have a choice between image and ISO before opening the modal on instance launch.

**Review by lorumic — COMMENTED** [review]
I like the way this is going. I think branching at the instance creation form stage is a good idea.
A few comments for some minor improvements follow.

**Comment by lorumic** [comment]
The 2 comments about the `appearance` and `caption` props apply also to the `SelectImageBtn` component.

File: src/pages/images/CustomIsoSelectorModal.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+interface Props {
+  onClose: () => void;
+  onSelect: (image: RemoteImage, type: string | null) => void;
+}
+
+export interface IsoImage {
+  name: string;
+  pool: string;
+}
```
[the remark this answers is not in the document]
Comment (edlerd) [reply src/pages/images/CustomIsoSelectorModal.tsx:15]: It is also used in `CustomIsoSelector`, but I'll put this in a type file and not have it next to the component.

File: src/pages/images/actions/SelectCustomIsoBtn.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  return (
+    <>
+      <Button
+        appearance={appearance}
+        onClick={openPortal}
+        type="button"
+        id="base-image"
+        className="image-btn"
+      >
+        <span>{caption}</span>
```
[the remark this answers is not in the document]
Comment (edlerd) [reply src/pages/images/actions/SelectCustomIsoBtn.tsx:30]: yes, the usages have changed so we can fixate the captions and appearances.

**Comment by edlerd** [comment]
Thanks for the comments @lorumic I cleaned up the code according to your suggestions and adjusted the modal sizing to be more stable.

File: src/pages/images/ImageSelectorModal.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (edlerd) [reply src/pages/images/ImageSelectorModal.tsx]: Yes, that should be less changy when switching the modals. I add some css for a more stable size.

[the remark this answers is not in the document]
Comment (lorumic) [reply src/pages/images/ImageSelectorModal.tsx]: I think the footer is supposed to stick to the bottom here, right?

image

File: src/pages/images/actions/UseCustomIsoBtn.tsx [hunk 5]
```diff
@@ -21,24 +19,21 @@ const SelectCustomIsoBtn: FC<Props> = ({ appearance, caption, onSelect }) => {
   return (
     <>
       <Button
-        appearance={appearance}
+        appearance="link"
         onClick={openPortal}
         type="button"
-        id="base-image"
+        id="select-custom-iso"
         className="image-btn"
```
Comment (lorumic) [inline src/pages/images/actions/UseCustomIsoBtn.tsx:26]: Is this class needed? Seems unused to me.

File: src/pages/images/actions/SelectImageBtn.tsx [hunk 6]
```diff
@@ -21,13 +19,13 @@ const SelectImageBtn: FC<Props> = ({ appearance, caption, onSelect }) => {
   return (
     <>
       <Button
-        appearance={appearance}
+        appearance="positive"
         onClick={openPortal}
         type="button"
-        id="base-image"
+        id="select-image"
         className="image-btn"
```
Comment (lorumic) [inline src/pages/images/actions/SelectImageBtn.tsx:26]: Same as above.

File: src/pages/images/ImageSelectorModal.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (edlerd) [reply src/pages/images/ImageSelectorModal.tsx]: right, I gave the table a wrapper with a min-height to have the footer pushed to the bottom of the modal.

**Review by lorumic — APPROVED** [review]
Thanks for applying the changes. It looks all good from a code and QA perspective.

**Review by piperdeck — CHANGES_REQUESTED** [review]
# 1

I can definitely see Michele's point about the empty ISO list creating an unnecessary click for the user. As discussed this morning, having a populated list here tested well in a user session yesterday. I think it's probably worth testing the empty list with a couple more users. I'm open to the idea of skipping this modal when the list is empty, but I'd like to see what users think of it first.

My vote is to leave it as-is for now, and revisit the topic with some user feedback in-hand.

## 2

This button can probably be set to the primary state, at least when the ISO list is empty.

Pasted image 20230908112827

**Comment by edlerd** [comment]
Thanks for the good suggestions @lorumic and @piperdeck . Making the upload primary in the empty ISO list makes a lot of sense, just pushed a change for that.

**Review by piperdeck — APPROVED** [review]
Looks good, approving!

**Pull request merged**